### DISCUSSION TITLE
Do push constants have alignment requirements?

### DISCUSSION THREAD

**Discussion by junglie85** [gfx-rs/wgpu#3373] [Q&A] [answered]
I am using a push constant to pass the model transform to my shader. This is a `glam::Mat4` as an array, which is uploaded as a `[f32; 16]` and has a size of 64. This works absolutely fine.

However, when I try uploading a `SpriteInstance` struct that is `bytemuck::Pod` and `bytemuck::Zeroable`, the screen is blank. This is only a size of 36 but AFAIK it meets the requirements for push constant data size (multiple of 4).

```
#[repr(C)]
#[derive(Debug, Clone, Copy)]
struct SpriteInstance {
    translation: [f32; 2],
    rotation: f32,
    scale: [f32; 2],
    tint: [f32; 4],
}

unsafe impl Pod for SpriteInstance {}
unsafe impl Zeroable for SpriteInstance {}

...

let sprite_push_constant_range = PushConstantRange {
        stages: ShaderStages::VERTEX,
        range: 0..size_of::<SpriteInstance>() as u32,
    };

...

 let sprite_pipeline_layout = device.create_pipeline_layout(&PipelineLayoutDescriptor {
        bind_group_layouts: &[
            &view_transforms_bind_group_layout,
            &sprite_bind_group_layout,
        ],
        push_constant_ranges: &[sprite_push_constant_range],
        label: Some("pipeline_layout::sprite"),
    });

...

rpass.set_push_constants(ShaderStages::VERTEX, 0, cast_slice(&[instance]));
```

And the shader:

```
struct SpriteInstance {
    translation: vec2<f32>,
    rotation: f32,
    scale: vec2<f32>,
    tint: vec4<f32>,
};

var<push_constant> instance: SpriteInstance;
```

FWIW, I also tried uploading each item separately, e.g.:

```
rpass.set_push_constants(ShaderStages::VERTEX, 0, cast_slice(&[instance.translation])); // offset 0, size 8
rpass.set_push_constants(ShaderStages::VERTEX, size_of::<[f32; 2]>() as u32, cast_slice(&[instance.rotation])); // offset 8, size 4
rpass.set_push_constants(ShaderStages::VERTEX, size_of::<[f32; 3]>() as u32, cast_slice(&[instance.scale])); // offset 12, size 8
rpass.set_push_constants(ShaderStages::VERTEX, size_of::<[f32; 5]>() as u32, cast_slice(&[instance.tint])); // offset 20, size 16
```

This neither worked with the push constant expecting the full struct nor with individual push constants.

I'm calculating the model matrix in the shader, but that works just fine (I use it and the `SpriteInstance` struct as instance data in another pipeline just fine). So I'm assuming it's to do with the push constants.

What's going on with the push constant that's preventing this from working?

-- Edit --

Having played a bit more and hardcoding some values in the shader, I've narrowed the issue down to the scale. Although the data size meets the requirements, so I ned to take alignment into account somehow?

**Comment by cwfitzgerald** [comment, marked as answer]
Push constants still follow std430 rules (i.e. wgsl block's rules for storage buffers).

- `translation` is at offset 0, size 8, alignment 8
- `rotation` is at offset 8, size 8, alignment 4
- `scale` is at offset _16_, size 8, alignment _8_
- `tint` is at offset _32_, size 16, alignment _16_

**Reply by junglie85** [reply]
Makes sense, thanks for confirming.
